Utilities/XrdAdaptor/src/XrdSource.cc

```diff
@@ -76,6 +76,110 @@ class DelayedClose : boost::noncopyable, public XrdCl::ResponseHandler
     std::string m_site;
 };
 
+
+/**
+ * A handler for querying a XrdCl::FileSystem object which is safe to be
+ * invoked from an XrdCl callback (that is, we don't need an available callback
+ * thread to timeout).
+ */
+class QueryAttrHandler : boost::noncopyable, public XrdCl::ResponseHandler
+{
+public:
+
+    virtual ~QueryAttrHandler() {}
+
+
+    static XrdCl::XRootDStatus query(XrdCl::FileSystem &fs, const std::string &attr, std::chrono::milliseconds timeout, std::string &result)
+    {
+        std::shared_ptr<QueryAttrHandler> handler(new QueryAttrHandler());
+        XrdCl::Buffer arg(attr.size());
+        arg.FromString(attr);
+
+        // On error or exception thrown, drop the reference to ourself
+        std::unique_ptr<QueryAttrHandler, std::function<void(QueryAttrHandler*)>> self_ref_guard(nullptr, [&](QueryAttrHandler*) {handler->m_self.reset();});
```

Comment:
Why not pull lines 99 to 110 within a block. That way the self_ref_guard usage is clearly deliniated.